azkaban-common/src/main/java/azkaban/executor/container/ContainerCleanupManager.java

```diff
@@ -325,9 +343,45 @@ private void cleanUpYarnApplicationsInCluster(Set<Integer> toBeCleanedContainers
       return;
     }
 
-    // todo: kill the apps asynchronously
+    // Use a fix thread pool to concurrently kill yarn apps
+    Map<String, Boolean> appsSuccessfulKilled = new ConcurrentHashMap<>();
+    aliveApplications.forEach(
+        app -> appsSuccessfulKilled.put(app.getApplicationId().toString(), false));
+
+    ExecutorService yarnKillThreadPool = Executors.newFixedThreadPool(this.yarnAppKillParallelism);
+    aliveApplications.forEach(app ->
```

Comment:
I think getApplication also should be in the async path

Reply:
The flow is this: main thread get all to-kill yarn app, then use a thread poll to kill each in async manner.
I cannot image what can we do making get-yarn-app async

Reply:
Yes as we discussed this is not the CM "main" thread, this thread is spawned every time CM schedules this method at fixed rate, so it won't jam the others or the "main" thread